app/src/main/java/com/lagradost/cloudstream3/ui/result/ResultViewModel2.kt

```diff
@@ -841,71 +904,243 @@ class ResultViewModel2 : ViewModel() {
+                context,
+                LibraryListType.FAVORITES,
+                CheckDuplicateData(
+                    name = response.name,
+                    year = response.year,
+                    syncData = response.syncData,
+                ),
+                getAllFavorites(),
+            ) { shouldContinue: Boolean, duplicateIds: List<Int?> ->
+                if (!shouldContinue) {
+                    statusChangedCallback?.invoke(null)
+                    return@checkAndWarnDuplicates
+                }
 
-            setFavoritesData(
-                currentId,
-                DataStoreHelper.FavoritesData(
+                if (duplicateIds.isNotEmpty()) {
+                    duplicateIds.forEach { duplicateId ->
+                        removeFavoritesData(duplicateId)
+                    }
+                }
+
+                val current = getFavoritesData(currentId)
+
+                setFavoritesData(
                     currentId,
-                    current?.favoritesTime ?: unixTimeMS,
-                    unixTimeMS,
-                    response.name,
-                    response.url,
-                    response.apiName,
-                    response.type,
-                    response.posterUrl,
-                    response.year
+                    DataStoreHelper.FavoritesData(
+                        current?.favoritesTime ?: unixTimeMS,
+                        currentId,
+                        unixTimeMS,
+                        response.name,
+                        response.url,
+                        response.apiName,
+                        response.type,
+                        response.posterUrl,
+                        response.year,
+                        response.syncData
+                    )
                 )
+
+                _favoriteStatus.postValue(true)
+
+                statusChangedCallback?.invoke(true)
+            }
+
+            return _favoriteStatus.value
+        }
+    }
+
+    @MainThread
+    private fun checkAndWarnDuplicates(
+        context: Context?,
+        listType: LibraryListType,
+        checkDuplicateData: CheckDuplicateData,
+        data: List<DataStoreHelper.LibrarySearchResponse>,
+        checkDuplicatesCallback: (shouldContinue: Boolean, duplicateIds: List<Int?>) -> Unit
+    ) {
+        val whitespaceRegex = "\\s+".toRegex()
+        fun normalizeString(input: String): String {
+            /**
+             * Trim the input string and replace consecutive spaces with a single space.
+             * This covers some edge-cases where the title does not match exactly across providers,
+             * and one provider has the title with an extra whitespace. This is minor enough that
+             * it should still match in this case.
+             */
+            return input.trim().replace(whitespaceRegex, " ")
+        }
+
+        val duplicateEntries = data.filter {
+            val syncData = checkDuplicateData.syncData
```

Comment:
move this outside the filter as well as all getImdbIdFromSyncData(librarySyncData) to reduce the amount of calls as this is quite a hot loop for many items this will improve the performance a bit

Reply:
that way we dont have to call normalizeString, hashmap get and getImdbIdFromSyncData getTMDbIdFromSyncData on every single item

Reply:
Thanks! Done.